### PR TITLE
feat: add SequentialTableTransformer

### LINKED ISSUE #802: Sequential table transformer
### Is your feature request related to a problem?

It's annoying to fit multiple transformers or sequentially transform a table using multiple transformers. This is also a very common operation.

### Desired solution

New class `SequentialTableTransformer`.

* Superclass: `InvertibleTableTransformer`
* Constructor parameters:
  * `transformers: list[TableTransformer]`
* `fit`:
  * Fit the first transformer with the given table
  * Transform the table with the first transformer and fit the second transformer with the transformed table
  * Repeat for the following transformers
* `transform`:
  * Transform the table with the first transformer
  * Transform the result with the second transformer
  * Repeat for the following transformers
* `inverse_transform`:
  * Raise an error if any transformer is not invertible
  * Inverse-transform the table with the last transformer
  * Inverse-transform the result with the second-to-last transformer
  * Repeat for the following transformers

### Possible alternatives (optional)

_No response_

### Screenshots (optional)

_No response_

### Additional Context (optional)

_No response_

### PR DESCRIPTION
**Pull request by RobinSen28**
Closes #802 

### Summary of Changes
Added ```SequentialTableTransformer``` class in transformers to transform tables sequentially using multiple transformers.
Added ```TransformerNotInvertibleError```, which gets raised when attempting to inverse transform a table using a non-invertible transformer.
Modified test helper ```assert_tables_equal``` to include more optional parameters.

### CODE REVIEW

**Comment by lars-reimann** [comment]
@RobinSen28 Please fix the merge conflicts, I'll review the PR once the automated checks pass afterwards.

File: tests/helpers/_assertions.py [hunk 1]
```diff
@@ -6,7 +6,15 @@
...
 
-def assert_tables_equal(table1: Table, table2: Table) -> None:
+def assert_tables_equal(
+    table1: Table,
+    table2: Table,
+    *,
+    ignore_column_order: bool = False,
+    ignore_row_order: bool = False,
+    ignore_types: bool = False,
+    check_exact: bool = False,
```
Comment (lars-reimann) [inline tests/helpers/_assertions.py:16]: Should either all be `ignore` or `check`. Having some options that enable and some that disable strict checking is confusing.

File: tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+from safeds.exceptions import TransformerNotFittedError, TransformerNotInvertibleError
+
+from tests.helpers import assert_tables_equal
+
+
+class TestInit:
+
+    def test_should_raise_value_error_on_empty_list(self) -> None:
+        with pytest.raises(ValueError, match=("transformers must contain at least 1 transformer")):
+            SequentialTableTransformer(transformers=[])  # type: ignore  # noqa: PGH003
```
Comment (lars-reimann) [inline tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py:21]: Ignore only the relavant mypy error: `type: ignore[...]` (replace ... by the error code. Then remove the `#noqa: PGH003`.

File: src/safeds/data/tabular/transformation/_sequential_table_transformer.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        self,
+        transformers: list[TableTransformer],
+        *,
+        column_names: str | list[str] | None = None,  # noqa: ARG002
+    ) -> None:
+        super().__init__(None)
+
+        # Check if transformers actually contains any transformers.
+        if transformers is None or len(transformers) == 0:
+            raise ValueError("transformers must contain at least 1 transformer")
```
Comment (lars-reimann) [inline src/safeds/data/tabular/transformation/_sequential_table_transformer.py:41]: The behavior of a sequential transformer that wraps an empty list is well defined, it's simply a no-op. We should, thus, get rid of the error and only issue a warning instead. See e.g. https://github.com/Safe-DS/Stdlib/blob/92622fb1bac5304c5127bf472e6630aceaf85c16/src/safeds/ml/classical/regression/_elastic_net_regressor.py#L42.

File: src/safeds/data/tabular/transformation/_sequential_table_transformer.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    ------
+    ValueError:
+        Raises a ValueError if the list of Transformers is None or contains no transformers.
+    """
+
+    def __init__(
+        self,
+        transformers: list[TableTransformer],
+        *,
+        column_names: str | list[str] | None = None,  # noqa: ARG002
```
Comment (lars-reimann) [inline src/safeds/data/tabular/transformation/_sequential_table_transformer.py:35]: Remove the unused parameter.

File: src/safeds/data/tabular/transformation/_sequential_table_transformer.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        fitted_transformers: list[TableTransformer] = []
+
+        for transformer in self._transformers:
+            fitted_transformer = transformer.fit(current_table)
+            fitted_transformers.append(fitted_transformer)
+            current_table = fitted_transformer.transform(current_table)
+
+        result: SequentialTableTransformer = SequentialTableTransformer(
+            transformers=fitted_transformers,
+            column_names=self._column_names,
```
Comment (lars-reimann) [inline src/safeds/data/tabular/transformation/_sequential_table_transformer.py:98]: Remove.

File: tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+
+class TestFit:
+    def test_should_raise_value_error_on_empty_table(self) -> None:
+        one_hot = OneHotEncoder()
+        imputer = SimpleImputer(SimpleImputer.Strategy.constant(0))
+        transformers = [one_hot, imputer]
```
Comment (lars-reimann) [inline tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py:28]: Tests should be as simple as possible. One transformer is enough to test that this error is raised.

File: tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+            },
+        )
+        sequential_table_transformer = SequentialTableTransformer(transformers)
+        with pytest.raises(
+            ValueError,
+            match=("The SequentialTableTransformer cannot be fitted because the table contains 0 rows."),
+        ):
+            sequential_table_transformer.fit(test_table)
+
+    def test_fit_does_not_change_original_transformer(self) -> None:
```
Comment (lars-reimann) [inline tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py:42]:
```suggestion
    def test_should_not_change_original_transformer(self) -> None:
```

The method under test is already specified by the wrapping class.

File: tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+                "col2": ["a", "b", "a"],
+            },
+        )
+        sequential_transformer = sequential_transformer.fit(test_table)
+        transformer = transformer.fit(test_table)
+        test_table_normal = transformer.transform(test_table)
+        test_table_sequential = sequential_transformer.transform(test_table)
+        assert_tables_equal(test_table_normal, test_table_sequential)
+
+    def test_transforms_correctly_with_multiple_transformers(self) -> None:
```
Comment (lars-reimann) [inline tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py:97]:
```suggestion
    def test_should_transform_with_multiple_transformers(self) -> None:
```

File: tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+        one_hot = one_hot.fit(test_table)
+        transormed_table_individual = one_hot.transform(test_table)
+        imputer = imputer.fit(transormed_table_individual)
+        transormed_table_individual = imputer.transform(transormed_table_individual)
+
+        assert_tables_equal(transfromed_table_sequential, transormed_table_individual)
+
+
+class TestIsFitted:
+    def test_should_return_false_before_fiting(self) -> None:
```
Comment (lars-reimann) [inline tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py:120]:
```suggestion
    def test_should_return_false_before_fitting(self) -> None:
```

File: tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+        one_hot = OneHotEncoder()
+        imputer = SimpleImputer(SimpleImputer.Strategy.constant(0))
+        transformers = [one_hot, imputer]
+        sequential_table_transformer = SequentialTableTransformer(transformers)
+        assert sequential_table_transformer.is_fitted is False
+
+    def test_should_return_true_after_fiting(self) -> None:
```
Comment (lars-reimann) [inline tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py:127]:
```suggestion
    def test_should_return_true_after_fitting(self) -> None:
```

File: tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+                "col1": [1, 2, None],
+                "col2": ["a", "b", "a"],
+            },
+        )
+        sequential_table_transformer = SequentialTableTransformer(transformers)
+        fitted_sequential_table_transformer = sequential_table_transformer.fit(test_table)
+        transfromed_table_sequential = fitted_sequential_table_transformer.transform(test_table)
+
+        one_hot = one_hot.fit(test_table)
+        transormed_table_individual = one_hot.transform(test_table)
```
Comment (lars-reimann) [inline tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py:112]:
```suggestion
        transformed_table_individual = one_hot.transform(test_table)
```

Use rename refactoring.

**Review by lars-reimann — CHANGES_REQUESTED** [review]

File: tests/helpers/_assertions.py (hunk 1 above)
Comment (xXstupidnameXx) [reply tests/helpers/_assertions.py]: Changed check_exact to ignore_float_imprecision and made it True by default

File: tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py (hunk 2 above)
Comment (xXstupidnameXx) [reply tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py]: replaced # type: ignore  # noqa: PGH003 with # type: ignore[attr-defined]

File: src/safeds/data/tabular/transformation/_sequential_table_transformer.py (hunk 3 above)
Comment (xXstupidnameXx) [reply src/safeds/data/tabular/transformation/_sequential_table_transformer.py]: Changed the code to only raise an error, when transformers in None and made it raise a warning, when the list is simply empty.

File: src/safeds/data/tabular/transformation/_sequential_table_transformer.py (hunk 5 above)
Comment (xXstupidnameXx) [reply src/safeds/data/tabular/transformation/_sequential_table_transformer.py]: Removed line column_names=self._column_names,

File: tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py (hunk 6 above)
Comment (xXstupidnameXx) [reply tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py]: changed the test to only use a SimpleImputer

File: tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py (hunk 7 above)
Comment (xXstupidnameXx) [reply tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py]: changed name of test to the one Suggested

File: tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py (hunk 9 above)
Comment (xXstupidnameXx) [reply tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py]: fixed typo

File: tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py (hunk 11 above)
Comment (xXstupidnameXx) [reply tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py]: fixed typo

File: tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py (hunk 10 above)
Comment (xXstupidnameXx) [reply tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py]: fixed typo

File: tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py (hunk 8 above)
Comment (xXstupidnameXx) [reply tests/safeds/data/tabular/transformation/test_sequential_table_transformer.py]: changed name of test to the one suggested

File: src/safeds/data/tabular/transformation/_sequential_table_transformer.py (hunk 4 above)
Comment (xXstupidnameXx) [reply src/safeds/data/tabular/transformation/_sequential_table_transformer.py]: Removed column_names

**Review by lars-reimann — CHANGES_REQUESTED** [review]

File: src/safeds/data/tabular/transformation/_sequential_table_transformer.py [hunk 12]
```diff
@@ -25,20 +26,25 @@ class SequentialTableTransformer(InvertibleTableTransformer):
...
     def __init__(
         self,
         transformers: list[TableTransformer],
-        *,
-        column_names: str | list[str] | None = None,  # noqa: ARG002
     ) -> None:
         super().__init__(None)
 
+        if transformers is None:
+            raise ValueError("transformers can't be None")
```
[the remark this answers is not in the document]
Comment (xXstupidnameXx) [reply src/safeds/data/tabular/transformation/_sequential_table_transformer.py]: removed the check

**Review by lars-reimann — APPROVED** [review]
👍

**Pull request merged**